new check: usage of deprecated variables or functions

arthurzam opened the pull request:
Resolves: https://github.com/pkgcore/pkgcheck/issues/373
Resolves: https://github.com/pkgcore/pkgcheck/issues/374

<details>
<summary>Example output</summary>

```
app-admin/gopass
  UnstableOnly: for arches: [ ppc64, x86 ], all versions are unstable: [ 1.12.6, 1.12.7, 1.13.0 ]
  DeprecatedEclassVariable: version 1.12.6: uses deprecated variable on line 8: EGO_SUM (no replacement)
  DeprecatedEclassFunction: version 1.12.6: uses deprecated function on line 494: go-module_set_globals (no replacement)
  PotentialStable: version 1.12.6: slot(0), stabled arch: [ amd64 ], potentials: [ ~ppc64, ~x86 ]
  DeprecatedEclassVariable: version 1.12.7: uses deprecated variable on line 8: EGO_SUM (no replacement)
  DeprecatedEclassFunction: version 1.12.7: uses deprecated function on line 498: go-module_set_globals (no replacement)
  RedundantVersion: version 1.12.7: slot(0) keywords are overshadowed by version: 1.13.0
  DeprecatedEclassVariable: version 1.13.0: uses deprecated variable on line 8: EGO_SUM (no replacement)
  DeprecatedEclassFunction: version 1.13.0: uses deprecated function on line 510: go-module_set_globals (no replacement)
  StableRequest: version 1.13.0: slot(0) no change in 146 days for unstable keyword: [ ~amd64 ]
```
</details>

---

TODO:

- [x] Add tests for this new check

thesamesam commented on src/pkgcheck/checks/eclass.py:
```diff
@@ -30,6 +30,40 @@ def desc(self):
+
+    @property
+    def desc(self):
+        if self.replacement is not True:
```
We're using a different pattern here than above (`is not None` vs `is not True`). Why not just `if self.replacement` always?

arthurzam replied:
We have 3 states:

1. `None` - not deprecated
2. `True` - no replacement string
3. other - the replacement string

I'm not sure what is the tag for eclass and why was it correct in that check. If I use `is not None` in my code, I get
```
DeprecatedEclassFunction: version 1.13.0: uses deprecated function: go-module_set_globals (migrate to True)
```

thesamesam replied:
Huh. OK, whatever. Thanks!

radhermit replied:
A better contextual explanation (off the top of my head but probably still relevant): pkgcore eclassdoc parsing provides three states:

1. False - not deprecated (the default value)
2. True - deprecated with no replacement (as specified in the eclass if the replacement value is 'None' or 'none')
3. string value - deprecated with a given value as the replacement

pkgcheck maps this into two states for the deprecated eclass result (as it ignores the non-deprecated state):

1. None - deprecated with no replacement
2. string value - deprecated with a given value as a replacement

What should be done for all related results is what sam said (and probably was just overlooked during writing/refactoring eclassdoc parsing initially). To do that for this check just copy what is done for eclasses and map non-string values (True) to None for the replacement argument.

thesamesam reviewed (commented):
Nice. Thanks for doing this so quickly. It seems reasonable to do the functions + variables together.

radhermit commented on src/pkgcheck/checks/eclass.py:
```diff
@@ -30,6 +30,40 @@ def desc(self):
+        return f'uses deprecated variable: {self.variable} ({replacement})'
+
+
+class DeprecatedEclassFunction(results.VersionResult, results.Warning):
```
This should be a LineResult for more context for the user.

radhermit commented on src/pkgcheck/checks/eclass.py:
```diff
@@ -30,6 +30,40 @@ def desc(self):
         return f'uses deprecated eclass: {self.eclass} ({replacement})'
 
 
+class DeprecatedEclassVariable(results.VersionResult, results.Warning):
```
This should be a LineResult for more context for the user.

radhermit commented on src/pkgcheck/checks/eclass.py:
```diff
@@ -98,6 +133,38 @@ def check_pre_inherits(self, pkg, inherits):
+
+        # scan for usage of @DEPRECATED variables
+        if deprecated_variables:
+            for node, _ in bash.var_assign_query.captures(pkg.tree.root_node):
```
I'd probably use `var_query` here instead of `var_assign_query` as I imagine some deprecated, eclass variables can be used without assigning to them (assuming this doesn't add false positives for whatever reason).

arthurzam commented:
@radhermit Thanks for the advises and suggestions, it is indeed better that way. I checked and didn't find false positives (at ::gentoo tree we don't have a lot of deprecated functions or variables).
@thesamesam I have split the check function into two separate functions, as there wasn't a lot gain from one, and now it is easier to read.

I have also added tests (taken some from the real `go-module.eclass` and did some manual changes to better check deprecated text)

thesamesam reviewed (approved):
Great work and thanks for doing this so quickly too.
